pkg/member/member_add.go

```diff
@@ -0,0 +1,264 @@
+package member
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/gardener/etcd-backup-restore/pkg/etcdutil"
+	"github.com/gardener/etcd-backup-restore/pkg/etcdutil/client"
+	"github.com/gardener/etcd-backup-restore/pkg/miscellaneous"
+	brtypes "github.com/gardener/etcd-backup-restore/pkg/types"
+	"github.com/sirupsen/logrus"
+	"go.etcd.io/etcd/clientv3"
+	"go.etcd.io/etcd/etcdserver/api/v3rpc/rpctypes"
+	"gopkg.in/yaml.v2"
+	"k8s.io/client-go/util/retry"
+)
+
+const (
+	// retryPeriod is the peroid after which an operation is retried
+	retryPeriod = 5 * time.Second
+
+	// etcdTimeout is timeout for etcd operations
+	etcdTimeout = 10 * time.Second
+)
+
+// AddMember interface defines the functionalities needed to add a member to the etcd cluster
+type AddMember interface {
+	// AddMemberAsLearner add a new member as a learner to the etcd cluster
+	AddMemberAsLearner(*logrus.Logger, *brtypes.EtcdConnectionConfig)
+
+	// IsMemberInCluster checks is the current members peer URL is already part of the etcd cluster
+	IsMemberInCluster(*logrus.Logger, *brtypes.EtcdConnectionConfig)
+
+	// PromoteMember promotes an etcd member from a learner to a voting member of the cluster. This will succeed only if its logs are caught up with the leader
+	PromoteMember(context.Context, *logrus.Entry, *brtypes.EtcdConnectionConfig)
+}
+
+// NewMember holds the configuration for the mechanism of adding a new member to the cluster.
+type NewMember struct {
+	clientFactory client.Factory
+}
+
+// NewMemberConfig returns new ExponentialBackoff.
+func NewMemberConfig(etcdConnConfig *brtypes.EtcdConnectionConfig) *NewMember {
+	etcdConn := *etcdConnConfig
+	svcEndpoint, err := miscellaneous.GetEtcdSvcEndpoint()
+	if svcEndpoint == "" || err != nil {
```

Comment:
I think you should return the err or print the error or distinguish between two cases `svcEndpoint == "" || err != nil`. Let’s suppose `err!=nil` then you are setting `svcEndpoint = "http://127.0.0.1:2379"` and then if we got some error due http instead of https then we thought it is some certificates issue which will give us false error.

Reply:
Yup, makes sense
Updated, and also updated `miscellaneous/GetEtcdSvcEndpoint()`
PTAL